### DISCUSSION TITLE
Allow explicit Windows SDK version

### DISCUSSION THREAD

**Discussion by reynoldsbd** [microsoft/vcpkg#23253] [Ideas]
****

**Is your feature request related to a problem? Please describe.**

For the sake of stability and reproducibility, my team prefers to build against an explicit version of the Windows SDK, e.g. 10.0.18362.0.

However, when loading its Visual Studio shell environment, vcpkg currently does not specify any particular version of the SDK. vcvarsall.bat will thus choose the latest available version. This limits our ability to control which version of the SDK is used when building vcpkg dependencies.

**Proposed solution**

Add a mechanism by which users may specify an explicit version of the Windows SDK to use. When calling vcvarsall.bat, vcpkg will pass the specified value using the `-winsdk=<ver>` argument.

I propose exposing this mechanism to the user by way of a new triplet variable, tentatively named `VCPKG_PLATFORM_SDK_VERSION`. 

**Describe alternatives you've considered**

It is technically possible to build with a specific Windows SDK version by simply uninstalling any newer versions of the SDK before running vcpkg. This is obviously cumbersome, particularly for users who may jump between projects with different SDK requirements.

My team is currently applying a patch to vcpkg-tool and bootstrapping manually. This patch is a very trivial hack which reads from an env var and injects the desired `-winsdk=<ver>` argument within _visualstudio.cpp_. Definitely not something that would be accepted upstream.

**Additional context**

I'm more than happy to submit a PR for this proposal. Just wanted to open this issue first to hash out the details and get feedback.

Unless there are objections, I plan on largely copying from microsoft/vcpkg-tool#199 which introduced `VCPKG_PLATFORM_TOOLSET_VERSION`.

**Comment by Hoikas** [comment]
I think you can do this by setting the `WindowsSDKVersion` environment variable and passing it through by setting `VCPKG_ENV_PASSTHROUGH` to `WindowsSDKVersion` in your triplet.

**Reply by reynoldsbd** [reply]
I wasn't able to make this work locally.

My understanding is that `WindowsSDKVersion` is an _output_ from vcvarsall.bat. So even if I set it to the desired value, it will be overwritten to the default (latest available) version once vcvarsall.bat is loaded.

**Reply by Neumann-A** [reply]
Set the env var in the triplet directly

**Comment by Cheney-W** [comment]
@BillyONeal @vicroms @ras0219-msft Could you please help take a look this issue?

**Comment by BillyONeal** [comment]
@Cheney-W I don't expect such a feature is going to make the cut line soon.

**Reply by reynoldsbd** [reply]
@BillyONeal can you elaborate on what you mean by "make the cut line"? I'm totally willing to implement this feature myself and submit a PR, but are you saying my PR would not be accepted?

**Reply by BillyONeal** [reply]
I mean "vcpkg engineers working on", not a statement about future PR acceptance. It seems like it would be useful but plumbing into all the build systems will not be a small undertaking.

**Reply by reynoldsbd** [reply]
Thanks for clarifying!

**Comment by solemnwarning** [comment]
Is there any progress on this?

I have a project which breaks with the latest Windows SDK, I can set the SDK used for the project and some of its libraries through VS (or msbuild), but as far as I can tell there's still no way to build the dependencies with vcpkg using the same version.
